ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/UserRightsChecker.scala

```
@@ -0,0 +1,69 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.auth
+
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.actor.Scheduler
+import com.daml.ledger.api.auth.interceptor.AuthorizationInterceptor
+import com.daml.ledger.participant.state.index.v2.UserManagementStore
+import com.daml.lf.data.Ref
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+private[auth] trait UserRightsChecker {
+  def schedule(userClaimsMismatchCallback: () => Unit): () => Unit
+}
+
+private[auth] final class UserRightsCheckerImpl(
+    lastUserRightsCheckTime: AtomicReference[Instant],
+    originalClaims: ClaimSet.Claims,
+    nowF: () => Instant,
+    userManagementStore: UserManagementStore,
+    userRightsCheckIntervalInSeconds: Int,
+    akkaScheduler: Scheduler,
+)(implicit ec: ExecutionContext)
+    extends UserRightsChecker {
+
+  def schedule(userClaimsMismatchCallback: () => Unit): () => Unit = {
+    if (originalClaims.resolvedFromUser) {
+      // Scheduling a task that periodically checks
+      // whether user rights state has changed.
+      // If user rights state has changed it aborts the stream by calling [[userClaimsMismatchCallback]]
+      val delay = userRightsCheckIntervalInSeconds.seconds
+      val userId = originalClaims.applicationId.fold[Ref.UserId](
```

this is no fails early, not sure if that is entirely right